src/common/ad_tx.c

```diff
@@ -608,3 +612,397 @@ ad_tx_redo_act_next(struct ad_tx *tx)
 	}
 	return &tx->tx_redo_act_pos->it_act;
 }
+
+static struct ad_tx *
+tx_get()
+{
+	return tls_tx;
+}
+
+static void
+tx_set(struct ad_tx *tx)
+{
+	tls_tx = tx;
+}
+
+static void
+tx_callback(struct ad_tx *tx)
+{
+	if (!tx->tx_stage_cb || tx->tx_layer != 0)
+		return;
+	tx->tx_stage_cb(ad_tx_stage(tx), tx->tx_stage_cb_arg);
+}
+
+static int
+tx_end(struct ad_tx *tx, int err)
+{
+	struct umem_tx	*utx;
+	int		 rc;
+
+	if (err == 0)
+		err = tx->tx_last_errno;
+	else
+		tx->tx_last_errno = err;
+
+	D_ASSERTF(tx->tx_layer >= 0, "TX "DF_U64", bad layer %d\n", ad_tx_id(tx), tx->tx_layer);
+	if (tx->tx_layer != 0)
+		return 0;
+
+	rc = ad_tx_end(tx, err);
+	if (err == 0) {
+		if (rc == 0) {
+			ad_tx_stage_set(tx, UMEM_STAGE_ONCOMMIT);
+		} else {
+			D_ERROR("ad_tx_end(%d) failed, "DF_RC"\n", err, DP_RC(rc));
+			tx->tx_last_errno = rc;
+			ad_tx_stage_set(tx, UMEM_STAGE_ONABORT);
+		}
+	} else {
+		D_ASSERT(rc != 0);
+		D_ASSERTF(ad_tx_stage(tx) == UMEM_STAGE_ONABORT,
+			  "TX "DF_U64", bad stage %d\n", ad_tx_id(tx), ad_tx_stage(tx));
+	}
+
+	tx_callback(tx);
+	if (tx_get() == tx)
```

Comment:
what does it mean if "tx_get() != tx"?

Reply:
Maybe we'd call this tx_set(NULL) before wal_commit() once the tx_end() is shared by both AD and BMEM.

Reply:
right, in umo_tx_commit() it set it to NULL already in this PR. here it to make sure all other cases will zero the tls (tx_abort for example).